lib/ecto/schema.ex

```diff
@@ -1139,9 +1196,66 @@ defmodule Ecto.Schema do
       # Update the order
       changeset = Repo.update!(changeset)
 
+  ## Inline embedded schema
+
+  The schema module can be defined inline in the parent schema in simple
+  cases:
+
+      defmodule Parent do
+        use Ecto.Schema
+
+        schema "parents" do
+          field :name, :string
+
+          embeds_one :children, Child do
+            field :name, :string
+            field :age,  :integer
+          end
+        end
+      end
+
+  Defining embedded schema in such a way will define a `Parent.Child` module
+  with the appropriate struct. In order to properly cast the embedded schema.
+  When casting the inline-defined embedded schemas you need to use the `:with`
+  option of `cast_embed/3` to provide the proper function to do the casting.
+  For example:
+
+      def changeset(schema, params) do
+        schema
+        |> cast(params, [:name])
+        |> cast_embed(:children, with: &child_changeset/2)
```

Comment:
What if in the casting code we checked if the function is defined and otherwise raised something along the lines of:

> the association module does not define a `changeset/2` function. You need to provide
> the function explicitly using the `:with` option.

This would probably cover more cases than just this and should cover this one as well.